pytest-6.0.1 does not find custom fixtures, test filename incorrect (just ..\..\)

```
============================= test session starts =============================
platform win32 -- Python 3.7.0, pytest-6.0.1, py-1.9.0, pluggy-0.13.1 -- c:\prog
ram files (x86)\python37-32\python.exe
cachedir: .pytest_cache
metadata: {'Python': '3.7.0', 'Platform': 'Windows-2008ServerR2-6.1.7601-SP1', '
Packages': {'pytest': '6.0.1', 'py': '1.9.0', 'pluggy': '0.13.1'}, 'Plugins': {'
html': '2.1.1', 'metadata': '1.10.0'}}
rootdir: C:\
plugins: html-2.1.1, metadata-1.10.0
collected 2 items

..\..\::test_install[uninstall_S0_0] ERROR (0 secs)

..\..\::test_install[ver_1.6.1_S0_1] ERROR (0 secs)


=================================== ERRORS ====================================
________________ ERROR at setup of test_install[uninstall_S0_0] _________________
file C:\install\tests\install_test.py, line 77
  @pytest.mark.parametrize("rel_info", release_dirs, ids=test_ids)
  def test_install(install_fixture, rel_info):
file C:\install\tests\install_test.py, line 26
  @pytest.fixture(scope='session')
  def module_init_fixture(session_fixture):
E       fixture 'session_fixture' not found
>       available fixtures: cache, capfd, capfdbinary, caplog, capsys, capsysbin
ary, doctest_namespace, extra, install_fixture, metadata, module_init_fixture, m
onkeypatch, pytestconfig, record_property, record_testsuite_property, record_xml
_attribute, recwarn, tmp_path, tmp_path_factory, tmpdir, tmpdir_factory
>       use 'pytest --fixtures [testpath]' for help on them.
```

- [ ] output of `pip list` from the virtual environment you are using
```
Package            Version
------------------ -------
atomicwrites       1.4.0
attrs              19.3.0
bcrypt             3.1.7
cffi               1.14.1
colorama           0.4.3
cryptography       3.0
importlib-metadata 1.7.0
iniconfig          1.0.1
json-flatten       0.1
more-itertools     8.4.0
packaging          20.4
paramiko           2.7.1
pip                10.0.1
pluggy             0.13.1
py                 1.9.0
pycparser          2.20
PyNaCl             1.4.0
pyparsing          2.4.7
pytest             6.0.1
pytest-html        2.1.1
pytest-metadata    1.10.0
scp                0.13.2
selenium           3.141.0
setuptools         39.0.1
six                1.15.0
toml               0.10.1
urllib3            1.25.10
zipp               3.1.0
```

 Python 3.7.0, pytest-6.0.1
OS Name:                   Microsoft Windows Server 2008 R2 Enterprise
OS Version:                6.1.7601 Service Pack 1 Build 7601

The same test work with pytest-5.3.2

Thank you.

I mimicked my fixture order with simple script. I am able to reproduce it. please find attached the script. Simply unzip under c:\ and run below command.
[test.zip](https://github.com/pytest-dev/pytest/files/5053060/test.zip)


pytest -s -v --show-capture=no \test\test\tests\dummy_test.py
============================= test session starts =============================
platform win32 -- Python 3.7.0, pytest-6.0.1, py-1.9.0, pluggy-0.13.1 -- c:\prog
ram files (x86)\python37-32\python.exe
cachedir: .pytest_cache
metadata: {'Python': '3.7.0', 'Platform': 'Windows-2008ServerR2-6.1.7601-SP1', '
Packages': {'pytest': '6.0.1', 'py': '1.9.0', 'pluggy': '0.13.1'}, 'Plugins': {'
html': '2.1.1', 'metadata': '1.10.0'}}
rootdir: C:\
plugins: html-2.1.1, metadata-1.10.0
collected 1 item

..\..\::test_dummy ERROR

=================================== ERRORS ====================================
________________________ ERROR at setup of test_dummy _________________________
file C:\test\test\tests\dummy_test.py, line 9
  def test_dummy(session_fixture2):
E       fixture 'session_fixture2' not found
>       available fixtures: cache, capfd, capfdbinary, caplog, capsys, capsysbin
ary, doctest_namespace, extra, metadata, module_fixture, monkeypatch, pytestconf
ig, record_property, record_testsuite_property, record_xml_attribute, recwarn, t
mp_path, tmp_path_factory, tmpdir, tmpdir_factory
>       use 'pytest --fixtures [testpath]' for help on them.

C:\test\test\tests\dummy_test.py:9
=========================== short test summary info ===========================
ERROR ..\..\::test_dummy
============================== 1 error in 0.03s ===============================

--- a/changelog/7628.bugfix.rst
+++ b/changelog/7628.bugfix.rst
@@ -1,0 +1,1 @@
+Fix test collection when a full path without a drive letter was passed to pytest on Windows (for example ``\projects\tests\test.py`` instead of ``c:\projects\tests\pytest.py``).

--- a/src/_pytest/main.py
+++ b/src/_pytest/main.py
@@ -549,7 +549,9 @@ class Session(nodes.FSCollector):
         self._initial_parts = []  # type: List[Tuple[py.path.local, List[str]]]
         self.items = items = []  # type: List[nodes.Item]
         for arg in args:
-            fspath, parts = self._parsearg(arg)
+            fspath, parts = resolve_collection_argument(
+                self.config.invocation_dir, arg, as_pypath=self.config.option.pyargs
+            )
             self._initial_parts.append((fspath, parts))
             initialpaths.append(fspath)
         self._initialpaths = frozenset(initialpaths)
@@ -673,37 +675,6 @@ class Session(nodes.FSCollector):
                 return
             yield from m
 
-    def _tryconvertpyarg(self, x: str) -> str:
-        """Convert a dotted module name to path."""
-        try:
-            spec = importlib.util.find_spec(x)
-        # AttributeError: looks like package module, but actually filename
-        # ImportError: module does not exist
-        # ValueError: not a module name
-        except (AttributeError, ImportError, ValueError):
-            return x
-        if spec is None or spec.origin is None or spec.origin == "namespace":
-            return x
-        elif spec.submodule_search_locations:
-            return os.path.dirname(spec.origin)
-        else:
-            return spec.origin
-
-    def _parsearg(self, arg: str) -> Tuple[py.path.local, List[str]]:
-        """Return (fspath, names) tuple after checking the file exists."""
-        strpath, *parts = str(arg).split("::")
-        if self.config.option.pyargs:
-            strpath = self._tryconvertpyarg(strpath)
-        fspath = Path(str(self.config.invocation_dir), strpath)
-        fspath = absolutepath(fspath)
-        if not fspath.exists():
-            if self.config.option.pyargs:
-                raise UsageError(
-                    "file or package not found: " + arg + " (missing __init__.py?)"
-                )
-            raise UsageError("file not found: " + arg)
-        return py.path.local(str(fspath)), parts
-
     def matchnodes(
         self, matching: Sequence[Union[nodes.Item, nodes.Collector]], names: List[str],
     ) -> Sequence[Union[nodes.Item, nodes.Collector]]:
@@ -770,3 +741,59 @@ class Session(nodes.FSCollector):
                 for subnode in rep.result:
                     yield from self.genitems(subnode)
             node.ihook.pytest_collectreport(report=rep)
+
+
+def search_pypath(module_name: str) -> str:
+    """Search sys.path for the given a dotted module name, and return its file system path."""
+    try:
+        spec = importlib.util.find_spec(module_name)
+    # AttributeError: looks like package module, but actually filename
+    # ImportError: module does not exist
+    # ValueError: not a module name
+    except (AttributeError, ImportError, ValueError):
+        return module_name
+    if spec is None or spec.origin is None or spec.origin == "namespace":
+        return module_name
+    elif spec.submodule_search_locations:
+        return os.path.dirname(spec.origin)
+    else:
+        return spec.origin
+
+
+def resolve_collection_argument(
+    invocation_dir: py.path.local, arg: str, *, as_pypath: bool = False
+) -> Tuple[py.path.local, List[str]]:
+    """Parse path arguments optionally containing selection parts and return (fspath, names).
+
+    Command-line arguments can point to files and/or directories, and optionally contain
+    parts for specific tests selection, for example:
+
+        "pkg/tests/test_foo.py::TestClass::test_foo"
+
+    This function ensures the path exists, and returns a tuple:
+
+        (py.path.path("/full/path/to/pkg/tests/test_foo.py"), ["TestClass", "test_foo"])
+
+    When as_pypath is True, expects that the command-line argument actually contains
+    module paths instead of file-system paths:
+
+        "pkg.tests.test_foo::TestClass::test_foo"
+
+    In which case we search sys.path for a matching module, and then return the *path* to the
+    found module.
+
+    If the path doesn't exist, raise UsageError.
+    """
+    strpath, *parts = str(arg).split("::")
+    if as_pypath:
+        strpath = search_pypath(strpath)
+    fspath = Path(str(invocation_dir), strpath)
+    fspath = absolutepath(fspath)
+    if not fspath.exists():
+        msg = (
+            "module or package not found: {arg} (missing __init__.py?)"
+            if as_pypath
+            else "file or directory not found: {arg}"
+        )
+        raise UsageError(msg.format(arg=arg))
+    return py.path.local(str(fspath)), parts

--- a/src/_pytest/main.py
+++ b/src/_pytest/main.py
@@ -31,6 +31,7 @@ from _pytest.config import UsageError
 from _pytest.config.argparsing import Parser
 from _pytest.fixtures import FixtureManager
 from _pytest.outcomes import exit
+from _pytest.pathlib import absolutepath
 from _pytest.pathlib import Path
 from _pytest.pathlib import visit
 from _pytest.reports import CollectReport
@@ -693,15 +694,15 @@ class Session(nodes.FSCollector):
         strpath, *parts = str(arg).split("::")
         if self.config.option.pyargs:
             strpath = self._tryconvertpyarg(strpath)
-        relpath = strpath.replace("/", os.sep)
-        fspath = self.config.invocation_dir.join(relpath, abs=True)
-        if not fspath.check():
+        fspath = Path(str(self.config.invocation_dir), strpath)
+        fspath = absolutepath(fspath)
+        if not fspath.exists():
             if self.config.option.pyargs:
                 raise UsageError(
                     "file or package not found: " + arg + " (missing __init__.py?)"
                 )
             raise UsageError("file not found: " + arg)
-        return (fspath, parts)
+        return py.path.local(str(fspath)), parts
 
     def matchnodes(
         self, matching: Sequence[Union[nodes.Item, nodes.Collector]], names: List[str],

--- a/testing/acceptance_test.py
+++ b/testing/acceptance_test.py
@@ -70,7 +70,7 @@ class TestGeneralUsage:
     def test_file_not_found(self, testdir):
         result = testdir.runpytest("asd")
         assert result.ret != 0
-        result.stderr.fnmatch_lines(["ERROR: file not found*asd"])
+        result.stderr.fnmatch_lines(["ERROR: file or directory not found: asd"])
 
     def test_file_not_found_unconfigure_issue143(self, testdir):
         testdir.makeconftest(
@@ -83,7 +83,7 @@ class TestGeneralUsage:
         )
         result = testdir.runpytest("-s", "asd")
         assert result.ret == ExitCode.USAGE_ERROR
-        result.stderr.fnmatch_lines(["ERROR: file not found*asd"])
+        result.stderr.fnmatch_lines(["ERROR: file or directory not found: asd"])
         result.stdout.fnmatch_lines(["*---configure", "*---unconfigure"])
 
     def test_config_preparse_plugin_option(self, testdir):
@@ -791,7 +791,7 @@ class TestInvocationVariants:
     def test_cmdline_python_package_not_exists(self, testdir):
         result = testdir.runpytest("--pyargs", "tpkgwhatv")
         assert result.ret
-        result.stderr.fnmatch_lines(["ERROR*file*or*package*not*found*"])
+        result.stderr.fnmatch_lines(["ERROR*module*or*package*not*found*"])
 
     @pytest.mark.xfail(reason="decide: feature or bug")
     def test_noclass_discovery_if_not_testcase(self, testdir):

--- a/testing/test_collection.py
+++ b/testing/test_collection.py
@@ -1426,3 +1426,42 @@ class TestImportModeImportlib:
                 "* 1 failed in *",
             ]
         )
+
+
+def test_module_full_path_without_drive(testdir):
+    """Collect and run test using full path except for the drive letter (#7628)
+
+    Passing a full path without a drive letter would trigger a bug in py.path.local
+    where it would keep the full path without the drive letter around, instead of resolving
+    to the full path, resulting in fixtures node ids not matching against test node ids correctly.
+    """
+    testdir.makepyfile(
+        **{
+            "project/conftest.py": """
+                import pytest
+                @pytest.fixture
+                def fix(): return 1
+            """,
+        }
+    )
+
+    testdir.makepyfile(
+        **{
+            "project/tests/dummy_test.py": """
+                def test(fix):
+                    assert fix == 1
+            """
+        }
+    )
+    fn = testdir.tmpdir.join("project/tests/dummy_test.py")
+    assert fn.isfile()
+
+    drive, path = os.path.splitdrive(str(fn))
+
+    result = testdir.runpytest(path, "-v")
+    result.stdout.fnmatch_lines(
+        [
+            os.path.join("project", "tests", "dummy_test.py") + "::test PASSED *",
+            "* 1 passed in *",
+        ]
+    )

--- a/testing/test_collection.py
+++ b/testing/test_collection.py
@@ -443,25 +443,6 @@ class TestCustomConftests:
 
 
 class TestSession:
-    def test_parsearg(self, testdir) -> None:
-        p = testdir.makepyfile("def test_func(): pass")
-        subdir = testdir.mkdir("sub")
-        subdir.ensure("__init__.py")
-        target = subdir.join(p.basename)
-        p.move(target)
-        subdir.chdir()
-        config = testdir.parseconfig(p.basename)
-        rcol = Session.from_config(config)
-        assert rcol.fspath == subdir
-        fspath, parts = rcol._parsearg(p.basename)
-
-        assert fspath == target
-        assert len(parts) == 0
-        fspath, parts = rcol._parsearg(p.basename + "::test_func")
-        assert fspath == target
-        assert parts[0] == "test_func"
-        assert len(parts) == 1
-
     def test_collect_topdir(self, testdir):
         p = testdir.makepyfile("def test_func(): pass")
         id = "::".join([p.basename, "test_func"])
@@ -1426,42 +1407,3 @@ class TestImportModeImportlib:
                 "* 1 failed in *",
             ]
         )
-
-
-def test_module_full_path_without_drive(testdir):
-    """Collect and run test using full path except for the drive letter (#7628)
-
-    Passing a full path without a drive letter would trigger a bug in py.path.local
-    where it would keep the full path without the drive letter around, instead of resolving
-    to the full path, resulting in fixtures node ids not matching against test node ids correctly.
-    """
-    testdir.makepyfile(
-        **{
-            "project/conftest.py": """
-                import pytest
-                @pytest.fixture
-                def fix(): return 1
-            """,
-        }
-    )
-
-    testdir.makepyfile(
-        **{
-            "project/tests/dummy_test.py": """
-                def test(fix):
-                    assert fix == 1
-            """
-        }
-    )
-    fn = testdir.tmpdir.join("project/tests/dummy_test.py")
-    assert fn.isfile()
-
-    drive, path = os.path.splitdrive(str(fn))
-
-    result = testdir.runpytest(path, "-v")
-    result.stdout.fnmatch_lines(
-        [
-            os.path.join("project", "tests", "dummy_test.py") + "::test PASSED *",
-            "* 1 passed in *",
-        ]
-    )

--- a/testing/test_main.py
+++ b/testing/test_main.py
@@ -1,8 +1,14 @@
 import argparse
+import os
+import re
 from typing import Optional
+
+import py.path
 
 import pytest
 from _pytest.config import ExitCode
+from _pytest.config import UsageError
+from _pytest.main import resolve_collection_argument
 from _pytest.main import validate_basetemp
 from _pytest.pytester import Testdir
 
@@ -98,3 +104,138 @@ def test_validate_basetemp_fails(tmp_path, basetemp, monkeypatch):
 def test_validate_basetemp_integration(testdir):
     result = testdir.runpytest("--basetemp=.")
     result.stderr.fnmatch_lines("*basetemp must not be*")
+
+
+class TestResolveCollectionArgument:
+    @pytest.fixture
+    def root(self, testdir):
+        testdir.syspathinsert(str(testdir.tmpdir / "src"))
+        testdir.chdir()
+
+        pkg = testdir.tmpdir.join("src/pkg").ensure_dir()
+        pkg.join("__init__.py").ensure(file=True)
+        pkg.join("test.py").ensure(file=True)
+        return testdir.tmpdir
+
+    def test_file(self, root):
+        """File and parts."""
+        assert resolve_collection_argument(root, "src/pkg/test.py") == (
+            root / "src/pkg/test.py",
+            [],
+        )
+        assert resolve_collection_argument(root, "src/pkg/test.py::") == (
+            root / "src/pkg/test.py",
+            [""],
+        )
+        assert resolve_collection_argument(root, "src/pkg/test.py::foo::bar") == (
+            root / "src/pkg/test.py",
+            ["foo", "bar"],
+        )
+        assert resolve_collection_argument(root, "src/pkg/test.py::foo::bar::") == (
+            root / "src/pkg/test.py",
+            ["foo", "bar", ""],
+        )
+
+    def test_dir(self, root):
+        """Directory and parts."""
+        assert resolve_collection_argument(root, "src/pkg") == (root / "src/pkg", [])
+        assert resolve_collection_argument(root, "src/pkg::") == (
+            root / "src/pkg",
+            [""],
+        )
+        assert resolve_collection_argument(root, "src/pkg::foo::bar") == (
+            root / "src/pkg",
+            ["foo", "bar"],
+        )
+        assert resolve_collection_argument(root, "src/pkg::foo::bar::") == (
+            root / "src/pkg",
+            ["foo", "bar", ""],
+        )
+
+    def test_pypath(self, root):
+        """Dotted name and parts."""
+        assert resolve_collection_argument(root, "pkg.test", as_pypath=True) == (
+            root / "src/pkg/test.py",
+            [],
+        )
+        assert resolve_collection_argument(
+            root, "pkg.test::foo::bar", as_pypath=True
+        ) == (root / "src/pkg/test.py", ["foo", "bar"],)
+        assert resolve_collection_argument(root, "pkg", as_pypath=True) == (
+            root / "src/pkg",
+            [],
+        )
+        assert resolve_collection_argument(root, "pkg::foo::bar", as_pypath=True) == (
+            root / "src/pkg",
+            ["foo", "bar"],
+        )
+
+    def test_does_not_exist(self, root):
+        """Given a file/module that does not exist raises UsageError."""
+        with pytest.raises(
+            UsageError, match=re.escape("file or directory not found: foobar")
+        ):
+            resolve_collection_argument(root, "foobar")
+
+        with pytest.raises(
+            UsageError,
+            match=re.escape(
+                "module or package not found: foobar (missing __init__.py?)"
+            ),
+        ):
+            resolve_collection_argument(root, "foobar", as_pypath=True)
+
+    def test_absolute_paths_are_resolved_correctly(self, root):
+        """Absolute paths resolve back to absolute paths."""
+        full_path = str(root / "src")
+        assert resolve_collection_argument(root, full_path) == (
+            py.path.local(os.path.abspath("src")),
+            [],
+        )
+
+        # ensure full paths given in the command-line without the drive letter resolve
+        # to the full path correctly (#7628)
+        drive, full_path_without_drive = os.path.splitdrive(full_path)
+        assert resolve_collection_argument(root, full_path_without_drive) == (
+            py.path.local(os.path.abspath("src")),
+            [],
+        )
+
+
+def test_module_full_path_without_drive(testdir):
+    """Collect and run test using full path except for the drive letter (#7628).
+
+    Passing a full path without a drive letter would trigger a bug in py.path.local
+    where it would keep the full path without the drive letter around, instead of resolving
+    to the full path, resulting in fixtures node ids not matching against test node ids correctly.
+    """
+    testdir.makepyfile(
+        **{
+            "project/conftest.py": """
+                import pytest
+                @pytest.fixture
+                def fix(): return 1
+            """,
+        }
+    )
+
+    testdir.makepyfile(
+        **{
+            "project/tests/dummy_test.py": """
+                def test(fix):
+                    assert fix == 1
+            """
+        }
+    )
+    fn = testdir.tmpdir.join("project/tests/dummy_test.py")
+    assert fn.isfile()
+
+    drive, path = os.path.splitdrive(str(fn))
+
+    result = testdir.runpytest(path, "-v")
+    result.stdout.fnmatch_lines(
+        [
+            os.path.join("project", "tests", "dummy_test.py") + "::test PASSED *",
+            "* 1 passed in *",
+        ]
+    )

--- a/testing/test_terminal.py
+++ b/testing/test_terminal.py
@@ -442,7 +442,9 @@ class TestCollectonly:
         have the items attribute."""
         result = testdir.runpytest("--collect-only", "uhm_missing_path")
         assert result.ret == 4
-        result.stderr.fnmatch_lines(["*ERROR: file not found*"])
+        result.stderr.fnmatch_lines(
+            ["*ERROR: file or directory not found: uhm_missing_path"]
+        )
 
     def test_collectonly_quiet(self, testdir):
         testdir.makepyfile("def test_foo(): pass")

